include/mrdox/Support/Handlebars.hpp

```diff
@@ -0,0 +1,1037 @@
+    renderDecorator(
+        Handlebars::Tag const& tag,
+        OutputRef &out,
+        std::string_view &templateText,
+        dom::Value const& data,
+        Handlebars::partials_map &inlinePartials,
+        dom::Object& private_data) const;
+
+    void
+    renderExpression(
+        Handlebars::Tag const& tag,
+        OutputRef &out,
+        std::string_view &templateText,
+        dom::Value const& data,
+        HandlebarsOptions const& opt,
+        dom::Object& private_data) const;
+
+    void
+    setupArgs(
+        std::string_view expression,
+        dom::Value const& data,
+        dom::Array &args,
+        HandlebarsCallback &cb) const;
+
+    dom::Value
+    evalExpr(
+        dom::Value const &context,
+        std::string_view expression) const;
+
+    helper_type const&
+    getHelper(std::string_view name, bool isBlock) const;
+};
+
+/** Determine if a value is truthy
+
+    A JSON value is truthy if it is a boolean and is true, a number and not
+    zero, or an non-empty string, array or object.
+
+    @param arg The value to test
+
+    @return True if the value is truthy, false otherwise
+
+ */
+MRDOX_DECL
+bool
+isTruthy(dom::Value const& arg);
+
+/** Determine if a value is empty
+
+    This is used by the built-in if and with helpers to control their
+    execution flow.
+
+    The Handlebars definition of empty is any of:
+
+    - Array with length 0
+    - falsy values other than 0
+
+    This is intended to match the Mustache Behaviour.
+
+    @param arg The value to test
+    @return True if the value is empty, false otherwise
+
+    @see https://mustache.github.io/mustache.5.html#Sections
+ */
+MRDOX_DECL
+bool
+isEmpty(dom::Value const& arg);
+
+/** Lookup a property in an object
+
+   @param data The object to look up the property in
+   @param path The path to the property to look up
+
+   @return The value of the property, or nullptr if the property does not exist
+ */
+MRDOX_DECL
+dom::Value
+lookupProperty(
+    dom::Value const &data,
+    std::string_view path);
+
+/** Stringify a value as JSON
+
+    This function converts a dom::Value to a string as if
+    JSON.stringify() had been called on it.
+
+    Recursive objects are identified.
+
+    @param v The value to stringify
+    @return The stringified value
+ */
+MRDOX_DECL
+std::string
+JSON_stringify(dom::Value const& value);
+
+namespace helpers {
+
+/** Register all the standard helpers into a Handlebars instance
+
+    @see https://github.com/handlebars-lang/handlebars.js/tree/master/lib/handlebars/helpers
+    @see https://handlebarsjs.com/guide/builtin-helpers.html
+
+    @param hbs The Handlebars instance to register the helpers into
+ */
+MRDOX_DECL
+void
+registerBuiltinHelpers(Handlebars& hbs);
+
+/// "if" helper function
+/**
+ * You can use the if helper to conditionally render a block. If its argument returns false, undefined, null,
+ * "", 0, or [], Handlebars will not render the block.
+ */
+MRDOX_DECL
+dom::Value
+if_fn(
```

Comment:
This is a little weird because you have C++ helpers, which can only be called from C++ code directly. To call them from scripts they need to go through the registration interface and the glue for the scripting language. And for C++ to call helpers that are scripted, they have to go through the registration interface. In other words, C++ calling a scripted helper might look like this (exposition only):
```
handlebars.callHelper( "myhelper", args, cb );
```

Aside from some open questions like how do we know the cb is the right one (shouldn't it be known to the handlebars instance because it is topmost on the "stack" ?)... now there are two different ways for C++ to call helpers. I'm not sure this is the right choice. It would be better if there was uniform syntax for invoking helpers from C++. 

Maybe since we are using C++20 you could make a constexpr version of my `callHelper` above and dispatch to the native helpers if the string matches? Otherwise use the scripting language's call operation to get the parameters onto the js/lua stack and invoke the script.

Reply:
> which can only be called from C++ code directly

I'm not sure I understand this problem. In theory these helpers *could* be called from C++ because they're valid, but they are not meant to be used directly. They are implementation details only meant to be called from templates. 

The templates use the `if_fn` because it's in a format in understands. But if I want an `if` statement in C++ or Javascript, I'll just use the `if` keyword without going through any `if_fn`. 

> now there are two different ways for C++ to call helpers

What two ways?

> uniform syntax for invoking helpers from C++.

I confirmed I'm not understanding any of this. 

There's no syntax to invoke helpers, unless you mean the helper signature handlebars expects, which is a single uniform syntax with two parameters: the array and the callback.

Reply:
Oh.. if the helpers are not meant to be called directly from C++ then ignore. Just make sure those things are private.

Reply:
What I did is I include functions such as

```
MRODX_DECL
void
registerAntoraHelpers(Handlebars& hbs);
```

that will register groups of helper functions that should be optional or at least have the logic separated. The individual helpers are just another option, but in any case they would always be registered directly with the engine. There's no point in calling them outside the templates to performs tasks because it's always going to be better to perform these tasks directly.

In the future, I was thinking of other similar groups of helper functions like `registerStringHelpers` with things like `join`, `split`, etc... because most applications will need these small helpers for string manipulation anyway.